libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Views/SFSDKWindowManager.m

```diff
@@ -0,0 +1,308 @@
+
+- (instancetype)init {
+    
+    self = [super init];
+    if (self) {
+        _namedWindows = [NSMapTable mapTableWithKeyOptions:NSMapTableStrongMemory
+                                              valueOptions:NSMapTableStrongMemory];
+        _delegates = [NSMutableOrderedSet orderedSet];
+    }
+    return self;
+    
+}
+
+- (SFSDKWindowContainer *)mainWindow {
+    if (![self.namedWindows objectForKey:kSFMainWindowKey]) {
+        [self setMainUIWindow:[SFApplicationHelper sharedApplication].windows[0]];
+    }
+    return [self.namedWindows objectForKey:kSFMainWindowKey];
+}
+
+- (void)setMainUIWindow:(UIWindow *) window {
+    SFSDKWindowContainer *container = [[SFSDKWindowContainer alloc] initWithWindow:window andName:kSFMainWindowKey];
+    
+    container.windowType = SFSDKWindowTypeMain;
+    container.windowLevel = window.windowLevel;
+    [container addDelegate:self];
+    _prevActiveWindow = container;
+    _currentWindow = container;
+    [self.namedWindows setObject:container forKey:kSFMainWindowKey];
+}
+
+- (SFSDKWindowContainer *)authWindow {
+    SFSDKWindowContainer *window = [self.namedWindows objectForKey:kSFLoginWindowKey];
+    if (!window) {
+        window = [self createAuthWindow];
+    }
+    return window;
+}
+
+- (SFSDKWindowContainer *)snapshotWindow {
+    SFSDKWindowContainer *window = [self.namedWindows objectForKey:kSFSnaphotWindowKey];
+    if (!window) {
+        window = [self createSnapshotWindow];
+    }
+    return window;
+}
+
+- (SFSDKWindowContainer *)passcodeWindow {
+    SFSDKWindowContainer *window = [self.namedWindows objectForKey:kSFPasscodeWindowKey];
+    if (!window) {
+        window = [self createPasscodeWindow];
+    }
+    return window;
+}
+
+- (SFSDKWindowContainer *)createNewNamedWindow:(NSString *)windowName {
+    SFSDKWindowContainer * container = nil;
+    if ( ![self isReservedName:windowName] ) {
+        UIWindow *window = [self createDefaultUIWindow];
+        container = [[SFSDKWindowContainer alloc] initWithWindow:window andName:windowName];
+        container.windowType = SFSDKWindowTypeOther;
+        [container addDelegate:self];
+        [self.namedWindows setObject:container forKey:windowName];
+    }
+    return container;
+}
+
+- (BOOL)isReservedName:(NSString *) windowName {
+    return ([windowName isEqualToString:kSFMainWindowKey] ||
+            [windowName isEqualToString:kSFLoginWindowKey] ||
+            [windowName isEqualToString:kSFPasscodeWindowKey] ||
+            [windowName isEqualToString:kSFSnaphotWindowKey]);
+    
+}
+
+- (BOOL)removeNamedWindow:(NSString *)windowName {
+    BOOL result = NO;
+    if (![self isReservedName:windowName]) {
+        [self.namedWindows removeObjectForKey:windowName];
+        result = YES;
+    }
+    return result;
+}
+
+- (SFSDKWindowContainer *)windowWithName:(NSString *)name {
+    return [self.namedWindows objectForKey:name];
+}
+
+- (void)bringToFront:(SFSDKWindowContainer *)windowContainer {
```

Comment:
Whatever form this method and its counterpart take—making a window visible or not—should a) be asynchronous, as with all UI display interactions, and b) ideally provide the option for animation, to control the presentation experience.